### ISSUE TITLE
[Feature Request] Outcome transformation that requires input parameters

### ISSUE THREAD

**Issue by pabloprf** [pytorch/botorch#1975]
How would you handle an **outcome transformation** that requires the (untransformed) `X` in the `forward` and `untransform` methods?

Is there something already built to handle this situation? 

If not, can you please provide a guideline of what would be required to make it happen? I imagine I would need to change the model class (e.g. `FixedNoiseGP`)  to take `train_X` at initialization. But then, where and how is the posterior un-transformation done?

@Balandat @saitcakmak 
Thanks!

**Comment by pabloprf** [comment]
In case it helps, the outcome transform that I need to implement constructs a scaling factor out of the inputs, and simply needs to multiply the posterior.

**Comment by Balandat** [comment]
Hmm interesting question. The `OutcomeTransform`s have been designed to only operate on the posterior / the outputs of the model. Changing that interface to somehow also accept the untransformed inputs could be quite painful; basically the only non-hacky way I can think of for this to work is if we changed the signature of `untransform` and `untransform_posterior` to also take in the (untransformed) `X`. But that would require changing the `forward` methods of all the models, and that seems quite a big change (given that this doesn't seem to be a very common situation).

@saitcakmak, @sdaulton, any additional thoughts on this?

**Comment by pabloprf** [comment]
Hi @Balandat, yes, I have been digging into the code and I agree that'd be a painful process.
Since most of the time I'm using a single model, I think I'll be able to make it work for my case, by redefining its `posterior` method. But it's possible that this is a feature that's not desirable to make it general in botorch, since it may be too specific for my application.

**Comment by pabloprf** [comment]
To get this working, I had to create my own outcome transformation class, which I created as a child of `botorch.models.transforms.outcome.Standardize`. Reason why I did this is because my outcome transformation is simply a factor `F` to multiply, which is the same as having `means = 0` and `stdvs = F`. This simplifies for me the construction of the `untransform` and `untransform_posterior` methods. I can construct the factor `F` from the untransformed inputs `X` and simply assign it to `stdvs` before calling the `super()` methods (and making sure that `self.training=False` so that it's not rewritten with the actual standard deviation). Both my `untransform` and `untransform_posterior` methods take `X` as input.

To make it work, I had to also create my own `ChainedOutcomeTransform` (because I want my transformation followed by standardization), and modify the methods to take also `X`. Then, I had to modify the `posterior` method of the `FixedNoiseGP` to also give the untransformed `X` to the outcome_transform call.

As @Balandat said, this was a bit painful, but the solution worked for my specific case. I hope this is useful to someone!
